src/crimson/common/chained_dispatchers.h

```diff
@@ -0,0 +1,46 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*- 
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <deque>
+#include <boost/intrusive/slist.hpp>
+
+#include "crimson/net/Dispatcher.h"
+#include "crimson/common/log.h"
+
+using Dispatcher = crimson::net::Dispatcher;
+
+// in existing Messenger, dispatchers are put into a chain as described by
+// chain-of-responsibility pattern. we could do the same to stop processing
+// the message once any of the dispatchers claims this message, and prevent
+// other dispatchers from reading it. but this change is more involved as
+// it requires changing the ms_ methods to return a bool. so as an intermediate 
+// solution, we are using an observer dispatcher to notify all the interested
+// or unintersted parties.
+class ChainedDispatchers {
```

Comment:
Um...actually, I thought, although only messengers own "chained dispatchers" for now, there may be other components that need it in the future, so I'm actually even planning to move Dispatcher to crimson/common. After all, dispatchers are not necessarily only related to messengers, right?:-)

Reply:
i think event notification and messages delivery is a service provided by messenger, and Dispatcher is the way how messenger offers this service.

> After all, dispatchers are not necessarily only related to messengers, right?:-)

agreed. but let's think this way, a post office boxes are not only related to the national postal service. it is an equipment installed everywhere. and is part of the machinery how the postal service offers the shipping and mailing service. so i consider it as a part of the postal service instead of the common municipal services or city services. one might argue, the common services covers municipal service as well as the postal service. but well, i'd think the "common" is like a catch all here. anything that is not big enough and is not "close" to a specific component, should go to "common".

Reply:
Agreed

Reply:
> ... so I'm actually even planning to move Dispatcher to crimson/common. After all, dispatchers are not necessarily only related to messengers, right?:-)

I think it's simpler and better to decouple if we follow the convention that `Dispatcher` is a set of listener interfaces that messenger exposes callbacks to higher-level components. The analogy is `PeeringState:: PeeringListener` and the concrete class `class PG : public PeeringState::PeeringListener`.